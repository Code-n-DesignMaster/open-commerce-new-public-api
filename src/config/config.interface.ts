import { GraphqlConfig } from '../modules/common/graphql.config';
import { AuthServiceConfig } from '../modules/auth/auth-service.config';

import { CommonConfig } from '../modules/common/common.config';
import { DaveBustersServiceConfig } from '../modules/dave-busters/dave-busters-service.config';
import { AuthorizationConfig } from '../modules/authorization/authorization.config';
import { CustomerConfig } from '../modules/customer/customer.config';
import { RateLimitDirectiveConfig } from '../modules/throttling-config/rate-limit-directive.config';
import { RollbarConfig } from '../modules/common/rollbar.config';
import { RedisConfig } from '../redis.config';
import { LogConfig } from '../log.config';
import { ThrottlingServiceConfig } from '../modules/throttling-config/throttling-service.config';

export interface ISchemaConfig {
  environment: string;
  port: number;

  common: CommonConfig;
  graphql: GraphqlConfig;
  rollbar: RollbarConfig;
  customer: CustomerConfig;
  authService: AuthServiceConfig;
  authorization: AuthorizationConfig;
  redis: RedisConfig;
  daveBusters: DaveBustersServiceConfig;
  rateLimitDirective: RateLimitDirectiveConfig;
  throttling: ThrottlingServiceConfig;
  log: LogConfig;
}
