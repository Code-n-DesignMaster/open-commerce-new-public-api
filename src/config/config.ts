export const config = {
  environment: {
    doc: 'The applicaton environment.',
    default: 'development',
    env: 'NODE_ENV',
  },
  port: {
    format: 'port',
    default: 3000,
    arg: 'port',
    env: 'PORT',
  },
  common: {
    minimumAppVersionAndroid: {
      doc: 'minimum version of android application',
      format: Number,
      env: 'MINIMUM_APP_VERSION_ANDROID',
      default: '329921',
    },
    minimumAppVersionIos: {
      doc: 'minimum version of ios application',
      format: Number,
      env: 'MINIMUM_APP_VERSION_IOS',
      default: '20191020.78',
    },
    apiVerificationUrl: {
      doc: 'Url to api verification service',
      format: 'url',
      env: 'API_VERIFICATION_URL',
      default: 'http://localhost:3050/graphql',
    },
  },
  customer: {
    isWhiteListForAutomatedQaEnabled: {
      doc:
        'Whitelist for automated tests enabled. Disables throttling on payment instrument create',
      format: Boolean,
      env: 'WHITELIST_FOR_AQA_TESTS_ENABLED',
      default: false,
    },
  },
  daveBusters: {
    toggleAppRatingPopup: {
      format: Boolean,
      env: 'DB__FEATURE_TOGGLE_APP_RATING_POPUP',
      default: false,
    },
    toggleRemovePowercardButton: {
      format: Boolean,
      env: 'DB__FEATURE_TOGGLE_REMOVE_POWERCARD_BUTTON',
      default: false,
    },
    payAtTableTipLabels: {
      format: Array,
      env: 'DB__PAY_AT_TABLE_TIP_LABELS',
      default: [0, 0, 0],
    },
  },
  rollbar: {
    accessToken: {
      format: String,
      env: 'ROLLBAR_ACCESS_TOKEN',
      default: undefined,
    },
    environment: {
      format: String,
      env: 'ROLLBAR_ENVIRONMENT',
      default: 'unknown',
    },
  },
  graphql: {
    typePaths: {
      doc:
        'The typePaths property indicates where the GraphQLModule should look for the GraphQL files.',
      format: Array,
      env: 'GQL_PATH',
      default: [
        'src/**/*.graphql',
        'node_modules/@open-commerce/graphql-schema/PUBLIC_API/schema-modules/transaction.graphqls',
      ],
    },
    debug: {
      doc: 'Debug mode',
      format: Boolean,
      env: 'GQL_DEBUG',
      default: false,
    },
    playground: {
      doc: 'Switch on playground',
      format: Boolean,
      env: 'GQL_PLAYGROUND',
      default: false,
    },
    resolverValidationOptions: {
      default: {
        requireResolversForResolveType: false,
      },
    },
  },
  authService: {
    inAuthTestMode: {
      doc:
        'Disables InAuth functionality for testing, returns mock InAuth responses',
      default: false,
      env: 'DB__INAUTH_TEST_MODE',
    },
  },
  authorization: {
    trustedIssuer: {
      doc: 'JWT issuer which is trusted',
      default: '',
      env: 'AUTH_SERVICE__TRUSTED_ISSUER',
    },
    jwksUri: {
      doc:
        'URI from which JWKS can be downloaded. Should be accessible by service',
      default: '',
      env: 'AUTH_SERVICE__JWKS_URI',
    },
  },
  throttling: {
    requests: {
      format: Number,
      env: 'PIN_THROTTLING_REQUESTS',
      default: 6,
    },
    timeInMinutes: {
      format: Number,
      env: 'PIN_THROTTLING_TIME_IN_MINUTES',
      default: 3,
    },
    throttledInMinutes: {
      format: Number,
      env: 'PIN_THROTTLING_THROTTLED_IN_MINUTES',
      default: 0,
    },
  },
  redis: {
    host: {
      doc: 'Connection host for Redis',
      format: String,
      env: 'REDIS_HOST',
      default: 'localhost',
    },
    port: {
      doc: 'Connection port for Redis',
      format: Number,
      env: 'REDIS_PORT',
      default: 6379,
    },
    db: {
      doc: 'Db number',
      format: Number,
      env: 'REDIS_DB',
      default: 0,
    },
    password: {
      doc: 'Redis connection password',
      format: String,
      env: 'REDIS_PASSWORD',
      default: undefined,
    },
    keyPrefix: {
      doc: 'Prefix for key',
      format: String,
      env: 'REDIS_PRIFIX',
      default: '',
    },
  },
  rateLimitDirective: {
    limit: {
      doc: 'number of requests per specific time',
      format: Number,
      env: 'RATE_LIMIT_NUMBER_OF_REQUEST',
      default: 3,
    },
    durationSeconds: {
      doc: 'time limit ',
      format: Number,
      env: 'RATE_LIMIT_TIME_LIMIT_SECONDS',
      // 6 hours
      default: 21600,
    },
    blockDurationSeconds: {
      doc: 'an amount of time user will be blocked after exceeding limit',
      format: Number,
      env: 'RATE_LIMIT_BLOCK_DURATION_SECONDS',
      default: 21600,
    },
    message: {
      doc: 'error message',
      format: String,
      env: 'RATE_LIMIT_ERROR_MESSAGE',
      default: 'Maximum number of requests reached. Please try again later',
    },
  },
  log: {
    level: {
      env: 'LOG_LEVEL',
      format: ['error', 'warn', 'info', 'debug'],
      default: 'info',
    },
    format: {
      env: 'LOG_FORMAT',
      format: ['console', 'json'],
      default: 'console',
    },
  },
};
