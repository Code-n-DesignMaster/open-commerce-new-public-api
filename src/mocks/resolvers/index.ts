import * as casual from 'casual';

export const Mutation = () => ({
  // your mutations
  acceptTermsAndConditions: () => {
    return casual.boolean;
  },
});

export const OCPhoneNumber = () => {
  return `+1${casual.phone.replace(/-/g, '')}`;
};

export const OCJSON = () => {
  return '{ foo: "bar" }';
};

export const OCDate = () => {
  return casual.date('YYYY-MM-DD');
};

export const OCDateTime = () => {
  return casual.date('YYYY-MM-DD') + 'T' + casual.time('HH:mm:ss') + 'Z';
};

export const OCURL = () => {
  return casual.url;
};

export const OCEmailAddress = () => {
  return casual.email;
};

export const OCTime = () => {
  return casual.date('HH:mm') + 'Z';
};
