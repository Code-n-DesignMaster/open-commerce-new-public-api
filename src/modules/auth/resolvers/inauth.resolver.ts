import { Args, Mutation, Query, Resolver } from '@nestjs/graphql';
import { InAuthDeviceRegistrationRequestInputDto } from '../dtos/inauth-device-registration-request-input.dto';
import { IInAuthDeviceRegistrationResponse } from '../interfaces/inauth-device-registration-response.interface';
import { Viewer } from '../../common/decorators/viewer.decorator';
import { IViewer } from '../../common/interfaces/viewer.interface';
import { AuthClientService } from '../services/auth-client.service';
import { IInAuthGetSigfilesResponse } from '../interfaces/inauth-get-sigfiles-response.interface';
import { InauthTrustScoreValueInputDto } from '../dtos/inauth-trust-score-value-input.dto';

@Resolver()
export class InAuthResolver {
  constructor(private readonly authService: AuthClientService) {}

  @Mutation('inauthDeviceTrustScoreThresholdUpdate')
  public async inauthDeviceTrustScoreThresholdUpdate(
    @Args('input') input: InauthTrustScoreValueInputDto,
  ): Promise<boolean> {
    return await this.authService.inauthDeviceTrustScoreThresholdUpdate(input);
  }

  @Mutation('inauthDeviceBlacklistRemove')
  public async inauthDeviceBlacklistRemove(
    @Args('deviceId') deviceId: string,
  ): Promise<boolean> {
    return await this.authService.inauthDeviceBlacklistRemove(deviceId);
  }

  @Mutation('inauthBlacklistedDeviceUnblockedStatusSet')
  public async inauthBlacklistedDeviceUnblockedStatusSet(
    @Args('deviceId') deviceId: string,
    @Args('unblockedStatus') unblockedStatus: boolean,
  ): Promise<boolean> {
    return await this.authService.inauthBlacklistedDeviceUnblockedStatusSet(
      deviceId,
      unblockedStatus,
    );
  }

  @Mutation()
  public async inAuthMobileDeviceRegister(
    @Args('input') input: InAuthDeviceRegistrationRequestInputDto,
    @Viewer() viewer: IViewer,
  ): Promise<IInAuthDeviceRegistrationResponse> {
    const { payload } = input;
    const { deviceId } = viewer;
    return await this.authService.inAuthMobileDeviceRegister(deviceId, payload);
  }

  @Query()
  public async inAuthGetSigfiles(
    @Viewer() viewer: IViewer,
  ): Promise<IInAuthGetSigfilesResponse> {
    const { deviceId, osType } = viewer;
    return await this.authService.inAuthGetSigfiles(deviceId, osType);
  }
}
