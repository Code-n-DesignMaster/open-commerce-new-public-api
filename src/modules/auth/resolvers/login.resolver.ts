import { Viewer } from './../../common/decorators/viewer.decorator';
import { Args, Mutation, Query, Resolver } from '@nestjs/graphql';
import { UseGuards, Inject } from '@nestjs/common';
import { LoggerService } from '@open-commerce/nestjs-logger';
import { ApolloError } from 'apollo-server-core';
import { OCNotAuthorizedError } from '../../../errors/OCNotAuthorizedError';
import { isEmpty } from 'lodash';
import assert = require('assert');
import { CustomerOtpRequestDto } from '../dtos/customer-otp-request.dto';
import { CustomerService } from '../../customer/customer.service';
import { AuthClientService } from '../services/auth-client.service';
import { PHONE_TYPE } from '../../customer/interfaces/phone-type.enum';
import { JwtAuthGuard } from '../../authorization/services/auth-guard.service';
import { Auth } from '../../common/decorators/auth.decorator';
import {
  THROTTLING_ERROR,
  ThrottlingService,
} from '@open-commerce/nestjs-throttling';
import { OCThrottlingError } from '../../../errors/OCThrottlingError';
import { IViewer } from '../../common/interfaces/viewer.interface';
import { CustomerLoginByOtpDto } from '../dtos/customer-login-by-otp.dto';
import {
  THROTTLING_CONFIG_SERVICE_TOKEN,
  ThrottlingConfig,
} from '../../throttling-config/throttling-config.constants';
import { IThrottlingConfigService } from '../../throttling-config/interfaces/throttling-config.service.interface';
import { IAuth } from '../../common/interfaces/auth.interface';

@Resolver('Auth')
export class LoginResolver {
  protected loggerContext: string = this.constructor.name;

  constructor(
    private customerService: CustomerService,
    private authService: AuthClientService,
    private logger: LoggerService,
    private throttlingService: ThrottlingService,
    @Inject(THROTTLING_CONFIG_SERVICE_TOKEN)
    private throttlingConfig: IThrottlingConfigService,
  ) {}

  @Query('loggedInTest')
  @UseGuards(JwtAuthGuard)
  public async loggedInTest(@Auth() auth: IAuth) {
    if (!auth.sub) {
      throw new OCNotAuthorizedError();
    }
    return {
      result: true,
    };
  }

  @Mutation('customerOTPRequest')
  public async customerOTPRequest(
    @Args('input') input: CustomerOtpRequestDto,
    @Viewer() viewer: IViewer,
  ) {
    try {
      this.logger.log('processing customerOTPRequest', this.loggerContext);

      const throttlingConfig = await this.throttlingConfig.get(
        ThrottlingConfig.CustomerOTPRequest,
      );
      await this.throttlingService.checkThrottledQueueAndIncrement(
        `customerOTPRequest_${input.identifier}`,
        {
          throttledInMinutes: throttlingConfig.blockDurationSeconds / 60,
          timeInMinutes: throttlingConfig.timeInSeconds / 60,
          requests: throttlingConfig.limit,
        },
      );

      const result = await this.authService.customerOTPRequest(input);

      this.logger.debug(
        {
          msg: 'processed customerOTPRequest',
          result: JSON.stringify(result),
        },
        this.loggerContext,
      );

      assert(
        !isEmpty(result.sub),
        Error(`JWT sub was not returned; login has likely failed`),
      );

      let customer = await this.customerService.retrieveBasicByIdentityId(
        result.sub,
      );

      if (!customer || !customer.uuid) {
        this.log({
          msg: 'no customer was retreived using result.sub',
          sub: result.sub,
        });

        customer = await this.customerService.createCustomer(null, {
          identityProviderId: result.sub,
          username: input.identifier,
          phoneNumbers: [
            {
              phoneNumber: input.identifier,
              phoneType: PHONE_TYPE.WORK,
            },
          ],
        });
      } else {
        this.log({
          msg: 'retreived customer using id',
          id: result.sub,
        });
      }

      // connect customer to device
      await this.authService.inauthCustomerToDeviceConnect(
        customer.id,
        viewer.deviceId,
      );

      return result;
    } catch (e) {
      this.handleError(e);
    }
  }

  @Mutation('customerLoginByOTP')
  public async customerLoginByOTP(
    @Args('input') input: CustomerLoginByOtpDto,
    @Args('inAuthRegistrationPayload') inAuthRegistrationPayload: string,
    @Viewer() viewer: IViewer,
  ) {
    // Only call device register if payload is present
    let inAuthData = null;

    if (inAuthRegistrationPayload) {
      inAuthData = await this.authService.inAuthMobileDeviceRegister(
        viewer.deviceId,
        inAuthRegistrationPayload,
      );
    }

    // Authenticate
    const authenticationResult = await this.authService.customerLoginByOTP(
      input,
      viewer.deviceId,
    );

    // Only include the inauth response if we have a registration payload
    const response = inAuthRegistrationPayload
      ? {
          ...authenticationResult,
          inAuthDeviceResponse: inAuthData.deviceResponse,
        }
      : {
          ...authenticationResult,
        };

    return response;
  }

  @Mutation('getAccessTokenByRefreshToken')
  public async getAccessTokenByRefreshToken(@Args('input') input) {
    return this.authService.getAccessTokenByRefreshToken(input);
  }

  private log(message: { msg: string; [key: string]: any }): void {
    this.logger.log(message, this.loggerContext);
  }

  private handleError(e) {
    // let's throw apollo error immediately
    if (e instanceof ApolloError) {
      throw e;
    }
    // let's rethrow errors from auth
    if (e.code && e.code.startsWith('OC_')) {
      throw new ApolloError(e.message, e.code);
    }

    if (e.code === THROTTLING_ERROR) {
      throw new OCThrottlingError(
        'Maximum number of code attempts reached. Please try again later.',
      );
    }

    throw e;
  }
}
