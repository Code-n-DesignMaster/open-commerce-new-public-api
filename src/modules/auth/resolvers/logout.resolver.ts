import { UseGuards } from '@nestjs/common';
import { Mutation, Resolver } from '@nestjs/graphql';
import { Auth } from '../../common/decorators/auth.decorator';
import { IAuth } from '../../common/interfaces/auth.interface';
import { JwtAuthGuard } from '../../authorization/services/auth-guard.service';
import { LogoutService } from '../services/logout.service';

@Resolver('Auth')
@UseGuards(JwtAuthGuard)
export class LogoutResolver {
  constructor(private logoutService: LogoutService) {}

  @Mutation('logout')
  public async logout(@Auth() auth: IAuth): Promise<boolean> {
    return this.logoutService.logout(auth);
  }
}
