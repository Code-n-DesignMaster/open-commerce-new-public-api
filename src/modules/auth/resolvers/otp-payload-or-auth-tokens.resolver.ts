import { Resolver, ResolveField } from '@nestjs/graphql';

@Resolver('CustomerOTPPayloadOrAuthTokens')
export class OtpPayloadOrAuthTokensResolver {
  @ResolveField()
  public async __resolveType(obj) {
    if ('tokens' in obj) {
      return 'AuthTokens';
    }
    return 'CustomerOTPPayload';
  }
}
