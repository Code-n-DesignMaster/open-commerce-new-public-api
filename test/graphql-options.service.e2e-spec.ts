import { Test, TestingModule } from '@nestjs/testing';
import { MocksService } from '@open-commerce/nestjs-mocks';
import { GraphQLModule, GqlModuleOptions } from '@nestjs/graphql';
import { GqlOptionsService } from '../src/modules/common/services/graphql-options.service';
import { ConfigModule } from '../src/config/config.module';
import * as resolvers from '../src/mocks/resolvers';
import { CONFIG } from '../src/config/config.constants';
import { MocksModule } from '@open-commerce/nestjs-mocks/dist';

// tslint:disable-next-line: no-var-requires
// const schema = require('../src/config/schema').schema;

describe('GraphQLOptionsService', () => {
  let moduleFixture: TestingModule;
  let app;
  let gqlSchemaOptions: GqlModuleOptions;

  beforeEach(async () => {
    moduleFixture = await Test.createTestingModule({
      imports: [
        ConfigModule,
        GraphQLModule.forRootAsync({
          imports: [
            ConfigModule,
            MocksModule.forRoot({
              typePath: './src/**/mocks/gqls/**/*.graphql',
              resolvers,
            }),
          ],
          useFactory: (
            config,
            mocksService: MocksService,
          ): GqlModuleOptions => {
            const optionsService = new GqlOptionsService(config, mocksService);
            gqlSchemaOptions = optionsService.createGqlOptions();
            return gqlSchemaOptions;
          },
          inject: [CONFIG, MocksService],
        }),
      ],
    }).compile();

    return null;
  });

  // it('should be constructed', () => {
  //   expect(gqlOptionsService).toBeDefined();
  // });
  //
  // it('has defined its services', () => {
  //   const services = (gqlOptionsService as any).services;
  //   expect(services.length).toBeGreaterThan(0);
  // });
  //
  // it('should create gql options', async () => {
  //   const gqlSchemaOptions = await gqlOptionsService.createGqlOptions();
  //
  //   expect(gqlSchemaOptions).toBeDefined();
  //   expect(gqlSchemaOptions).toMatchSnapshot();
  //
  //   app = await moduleFixture.createNestApplication();
  //   return await app.init();
  // });

  it('should call transformSchema', async () => {
    // It solution doesn't work!
    // better to discuss it.
    const spy = jest.spyOn(gqlSchemaOptions, 'transformSchema');

    app = await moduleFixture.createNestApplication();
    await app.init();
    return expect(spy).toHaveBeenCalled();
  });

  it.skip('should create gql options', async () => {
    const spy = jest.spyOn(gqlSchemaOptions, 'transformSchema');

    app = await moduleFixture.createNestApplication();
    await app.init();
    return expect(spy.mock.calls[0]).toMatchSnapshot();
  });

  it.skip('should create gql options', async () => {
    const spy = jest.spyOn(gqlSchemaOptions, 'transformSchema');

    app = await moduleFixture.createNestApplication();
    await app.init();
    return expect(spy.mock.results[0].value).resolves.toMatchSnapshot();
  });

  afterEach(() => {
    // dirty hack, should remove it.
    if (app) {
      app.close();
    }
  });
});
