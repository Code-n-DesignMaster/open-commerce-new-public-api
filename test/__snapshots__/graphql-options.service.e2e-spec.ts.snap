// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQLOptionsService should create gql options 1`] = `
Array [
  GraphQLSchema {
    "__allowedLegacyNames": Array [],
    "__validationErrors": undefined,
    "_directives": Array [
      "@skip",
      "@include",
      "@deprecated",
    ],
    "_implementations": Object {
      "Node": Array [
        "Site",
        "Customer",
        "Transaction",
      ],
    },
    "_mutationType": "Mutation",
    "_possibleTypeMap": undefined,
    "_queryType": "Query",
    "_subscriptionType": null,
    "_typeMap": Object {
      "Attribute": "Attribute",
      "AttributeInput": "AttributeInput",
      "AuthTokens": "AuthTokens",
      "Boolean": "Boolean",
      "CarWashOption": "CarWashOption",
      "CompletedResultEnum": "CompletedResultEnum",
      "Customer": "Customer",
      "CustomerInput": "CustomerInput",
      "CustomerOTPPayload": "CustomerOTPPayload",
      "CustomerOTPRequest": "CustomerOTPRequest",
      "Date": "Date",
      "DateTime": "DateTime",
      "Decimal": "Decimal",
      "Float": "Float",
      "FuelProduct": "FuelProduct",
      "ID": "ID",
      "Int": "Int",
      "LoyaltyLevelEnum": "LoyaltyLevelEnum",
      "Mutation": "Mutation",
      "Node": "Node",
      "OTP": "OTP",
      "OTPChannel": "OTPChannel",
      "OTPMode": "OTPMode",
      "PageInfo": "PageInfo",
      "PaymentLevelEnum": "PaymentLevelEnum",
      "PaymentLocationEnum": "PaymentLocationEnum",
      "PaymentMethod": "PaymentMethod",
      "ProcessingStateEnum": "ProcessingStateEnum",
      "Pump": "Pump",
      "PumpStatusEnum": "PumpStatusEnum",
      "Query": "Query",
      "Receipt": "Receipt",
      "SaleItem": "SaleItem",
      "ServiceLevelEnum": "ServiceLevelEnum",
      "Site": "Site",
      "SiteStatusEnum": "SiteStatusEnum",
      "String": "String",
      "Time": "Time",
      "Tokens": "Tokens",
      "Transaction": "Transaction",
      "TransactionBeginAtPumpInput": "TransactionBeginAtPumpInput",
      "TransactionConnection": "TransactionConnection",
      "TransactionEdge": "TransactionEdge",
      "TransactionFilter": "TransactionFilter",
      "__Directive": "__Directive",
      "__DirectiveLocation": "__DirectiveLocation",
      "__EnumValue": "__EnumValue",
      "__Field": "__Field",
      "__InputValue": "__InputValue",
      "__Schema": "__Schema",
      "__Type": "__Type",
      "__TypeKind": "__TypeKind",
    },
    "astNode": Object {
      "directives": Array [],
      "kind": "SchemaDefinition",
      "operationTypes": Array [
        Object {
          "kind": "OperationTypeDefinition",
          "operation": "query",
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Query",
            },
          },
        },
        Object {
          "kind": "OperationTypeDefinition",
          "operation": "mutation",
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Mutation",
            },
          },
        },
      ],
    },
    "extensionASTNodes": Array [],
  },
]
`;

exports[`GraphQLOptionsService should create gql options 2`] = `
GraphQLSchema {
  "__allowedLegacyNames": Array [],
  "__validationErrors": Array [],
  "_directives": Array [
    "@include",
    "@skip",
    "@deprecated",
  ],
  "_implementations": Object {
    "CustomerOTPPayloadOrAuthTokens": Array [],
    "Node": Array [
      "Customer",
      "Transaction",
      "Site",
    ],
  },
  "_mutationType": "Mutation",
  "_possibleTypeMap": undefined,
  "_queryType": "Query",
  "_subscriptionType": null,
  "_typeMap": Object {
    "AccessTokenByRefreshTokenInput": "AccessTokenByRefreshTokenInput",
    "AccessTokenByRefreshTokenInputType": "AccessTokenByRefreshTokenInputType",
    "AddPowercardFunds": "AddPowercardFunds",
    "Address": "Address",
    "AttractionChip": "AttractionChip",
    "Attribute": "Attribute",
    "AttributeInput": "AttributeInput",
    "AuthChallenge": "AuthChallenge",
    "AuthChallengeType": "AuthChallengeType",
    "AuthPayloadType": "AuthPayloadType",
    "AuthTokens": "AuthTokens",
    "Boolean": "Boolean",
    "Brand": "Brand",
    "BrandAgreement": "BrandAgreement",
    "BrandFilter": "BrandFilter",
    "CUSTOMER_REQUEST_TYPE": "CUSTOMER_REQUEST_TYPE",
    "CUSTOMER_STATUS": "CUSTOMER_STATUS",
    "CarWashOption": "CarWashOption",
    "CheckIntoLocation": "CheckIntoLocation",
    "ClientToken": "ClientToken",
    "CompletedResultEnum": "CompletedResultEnum",
    "Customer": "Customer",
    "CustomerAgreement": "CustomerAgreement",
    "CustomerContact": "CustomerContact",
    "CustomerCreate": "CustomerCreate",
    "CustomerCreateAddress": "CustomerCreateAddress",
    "CustomerCreateDemographics": "CustomerCreateDemographics",
    "CustomerCreatePersonName": "CustomerCreatePersonName",
    "CustomerCreatePhoneNumber": "CustomerCreatePhoneNumber",
    "CustomerCreateVerify": "CustomerCreateVerify",
    "CustomerDevice": "CustomerDevice",
    "CustomerDeviceCreate": "CustomerDeviceCreate",
    "CustomerDeviceFilter": "CustomerDeviceFilter",
    "CustomerDeviceUpdate": "CustomerDeviceUpdate",
    "CustomerFeedback": "CustomerFeedback",
    "CustomerInput": "CustomerInput",
    "CustomerLoginAttempt": "CustomerLoginAttempt",
    "CustomerLoginAttemptConnection": "CustomerLoginAttemptConnection",
    "CustomerLoginAttemptEdge": "CustomerLoginAttemptEdge",
    "CustomerLoginAttemptFilter": "CustomerLoginAttemptFilter",
    "CustomerLoyaltyInstrumentCreate": "CustomerLoyaltyInstrumentCreate",
    "CustomerLoyaltyInstrumentUpdate": "CustomerLoyaltyInstrumentUpdate",
    "CustomerOTPPayload": "CustomerOTPPayload",
    "CustomerOTPPayloadOrAuthTokens": "CustomerOTPPayloadOrAuthTokens",
    "CustomerOTPRequest": "CustomerOTPRequest",
    "CustomerOTPVerify": "CustomerOTPVerify",
    "CustomerPassword": "CustomerPassword",
    "CustomerPasswordChange": "CustomerPasswordChange",
    "CustomerPasswordCredentials": "CustomerPasswordCredentials",
    "CustomerPasswordReset": "CustomerPasswordReset",
    "CustomerPasswordResetByOTPVerify": "CustomerPasswordResetByOTPVerify",
    "CustomerPasswordResetPaymentInstrumentVerify": "CustomerPasswordResetPaymentInstrumentVerify",
    "CustomerPasswordSet": "CustomerPasswordSet",
    "CustomerPaymentInstrumentCreate": "CustomerPaymentInstrumentCreate",
    "CustomerPaymentInstrumentUpdate": "CustomerPaymentInstrumentUpdate",
    "CustomerPaymentInstrumentVerify": "CustomerPaymentInstrumentVerify",
    "CustomerPinCode": "CustomerPinCode",
    "CustomerPinCodeChange": "CustomerPinCodeChange",
    "CustomerPinCodeCredentials": "CustomerPinCodeCredentials",
    "CustomerPinCodeSet": "CustomerPinCodeSet",
    "CustomerPinCodeVerify": "CustomerPinCodeVerify",
    "CustomerUpdate": "CustomerUpdate",
    "CustomerUpdateAddress": "CustomerUpdateAddress",
    "CustomerUpdateDemographics": "CustomerUpdateDemographics",
    "CustomerUpdatePersonName": "CustomerUpdatePersonName",
    "CustomerUpdatePhoneNumber": "CustomerUpdatePhoneNumber",
    "CustomerUsername": "CustomerUsername",
    "Date": "Date",
    "DateTime": "DateTime",
    "Decimal": "Decimal",
    "Demographics": "Demographics",
    "FUEL_PUMP_STATUS": "FUEL_PUMP_STATUS",
    "Float": "Float",
    "FuelProduct": "FuelProduct",
    "FuelReceipt": "FuelReceipt",
    "GeoLocation": "GeoLocation",
    "GeoLocationFilter": "GeoLocationFilter",
    "HARDWARE_VENDOR": "HARDWARE_VENDOR",
    "HoursOfOperation": "HoursOfOperation",
    "ID": "ID",
    "Int": "Int",
    "JSON": "JSON",
    "LOCATION_RESOURCE_STATUS": "LOCATION_RESOURCE_STATUS",
    "LOCATION_RESOURCE_TYPE": "LOCATION_RESOURCE_TYPE",
    "LOYALTY_INSTRUMENT_STATUS": "LOYALTY_INSTRUMENT_STATUS",
    "LOYALTY_LEVEL": "LOYALTY_LEVEL",
    "Location": "Location",
    "LocationAttribute": "LocationAttribute",
    "LocationAttributeFilter": "LocationAttributeFilter",
    "LocationConnection": "LocationConnection",
    "LocationEdge": "LocationEdge",
    "LocationFilter": "LocationFilter",
    "LocationResource": "LocationResource",
    "LocationResourceFilter": "LocationResourceFilter",
    "LoyaltyInstrument": "LoyaltyInstrument",
    "LoyaltyInstrumentFilter": "LoyaltyInstrumentFilter",
    "LoyaltyLevelEnum": "LoyaltyLevelEnum",
    "Mutation": "Mutation",
    "NewAttractionChip": "NewAttractionChip",
    "NewGeoLocation": "NewGeoLocation",
    "NewPowercard": "NewPowercard",
    "NewPowercardImage": "NewPowercardImage",
    "NewVirtualPowercard": "NewVirtualPowercard",
    "Node": "Node",
    "OTP": "OTP",
    "OTPChannel": "OTPChannel",
    "OTPMode": "OTPMode",
    "OTP_CHANNEL": "OTP_CHANNEL",
    "OTP_MODE": "OTP_MODE",
    "OpenCommerceFeature": "OpenCommerceFeature",
    "OpenCommerceFeatureFilter": "OpenCommerceFeatureFilter",
    "OtpInput": "OtpInput",
    "PAYMENT_INSTRUMENT_STATUS": "PAYMENT_INSTRUMENT_STATUS",
    "PAYMENT_INSTRUMENT_TYPE": "PAYMENT_INSTRUMENT_TYPE",
    "PAYMENT_LEVEL": "PAYMENT_LEVEL",
    "PHONE_TYPE": "PHONE_TYPE",
    "POWERCARD_CARD_TYPE": "POWERCARD_CARD_TYPE",
    "POWERCARD_STATUS_TYPE": "POWERCARD_STATUS_TYPE",
    "PageInfo": "PageInfo",
    "PaymentInstrument": "PaymentInstrument",
    "PaymentLevelEnum": "PaymentLevelEnum",
    "PaymentLocationEnum": "PaymentLocationEnum",
    "PaymentMethod": "PaymentMethod",
    "PersonName": "PersonName",
    "PhoneNumber": "PhoneNumber",
    "Powercard": "Powercard",
    "PowercardFilter": "PowercardFilter",
    "PowercardImage": "PowercardImage",
    "PowercardImageFilter": "PowercardImageFilter",
    "ProcessingStateEnum": "ProcessingStateEnum",
    "Pump": "Pump",
    "PumpStatusEnum": "PumpStatusEnum",
    "Query": "Query",
    "RESOURCE_ITEM_TYPE": "RESOURCE_ITEM_TYPE",
    "RateCard": "RateCard",
    "Receipt": "Receipt",
    "ReceiptLine": "ReceiptLine",
    "ResourceItem": "ResourceItem",
    "ResourceItemConnection": "ResourceItemConnection",
    "ResourceItemEdge": "ResourceItemEdge",
    "ResourceItemFilter": "ResourceItemFilter",
    "ResourceItemPrice": "ResourceItemPrice",
    "Response": "Response",
    "STAC": "STAC",
    "SaleItem": "SaleItem",
    "ServiceLevelEnum": "ServiceLevelEnum",
    "SignUpInput": "SignUpInput",
    "Site": "Site",
    "SiteStatusEnum": "SiteStatusEnum",
    "StartTransaction": "StartTransaction",
    "String": "String",
    "TRANSACTION_STATUS": "TRANSACTION_STATUS",
    "TRANSACTION_TYPE": "TRANSACTION_TYPE",
    "Tenant": "Tenant",
    "Time": "Time",
    "Tokens": "Tokens",
    "Transaction": "Transaction",
    "TransactionBeginAtPumpInput": "TransactionBeginAtPumpInput",
    "TransactionConnection": "TransactionConnection",
    "TransactionEdge": "TransactionEdge",
    "TransactionFilter": "TransactionFilter",
    "TransactionPaymentInfo": "TransactionPaymentInfo",
    "TransactionPaymentInfoFilter": "TransactionPaymentInfoFilter",
    "UpdateGeoLocation": "UpdateGeoLocation",
    "UpdatePowercardAttributes": "UpdatePowercardAttributes",
    "UserAttributes": "UserAttributes",
    "UserLoginInput": "UserLoginInput",
    "WEEKDAY": "WEEKDAY",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
  },
  "astNode": undefined,
  "extensionASTNodes": undefined,
  "transforms": Array [
    FilterRootFields {
      "transformer": TransformRootFields {
        "transform": [Function],
      },
    },
  ],
}
`;
